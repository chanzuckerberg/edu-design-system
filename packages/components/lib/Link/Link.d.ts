import React, { ReactNode } from "react";
import { ClickableStyleProps } from "../ClickableStyle";
declare type LinkHTMLElementProps = Omit<React.AnchorHTMLAttributes<HTMLAnchorElement>, "disabled">;
export declare type LinkProps = LinkHTMLElementProps & {
    /**
     * The link contents or label.
     */
    children: ReactNode;
    color?: ClickableStyleProps<"a">["color"];
    "data-testid"?: string;
    size?: ClickableStyleProps<"a">["size"];
    variant?: ClickableStyleProps<"a">["variant"];
};
/**
 * Component for making anchor tags.
 *
 * This component is called Link because it should be used to make `<a>` elements;
 * however, it can be styled to look like a button.
 *
 * If you need to style a `<button>` element to look like a link, please use the `Button` component.
 * If you need to style a different element or component (e.g. `Link` from `react-router`) to
 * look like a button or link, you can use the `ClickableStyle` component.
 *
 * In terms of the look and feel of the component in the UI, the `Button`, and `Link`, and `ClickableStyle`
 * components are exactly the same.
 */
declare const Link: React.ForwardRefExoticComponent<LinkHTMLElementProps & {
    /**
     * The link contents or label.
     */
    children: ReactNode;
    color?: "alert" | "brand" | "neutral" | "success" | "warning" | undefined;
    "data-testid"?: string | undefined;
    size?: "small" | "medium" | "large" | undefined;
    variant?: "link" | "flat" | "outline" | "plain" | undefined;
} & React.RefAttributes<HTMLAnchorElement>>;
export default Link;
