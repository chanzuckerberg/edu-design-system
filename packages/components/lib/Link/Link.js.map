{"version":3,"file":"Link.js","sourceRoot":"","sources":["../../src/Link/Link.tsx"],"names":[],"mappings":";;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;AAAA,+CAAqD;AACrD,uEAAwE;AAkBxE;;;;;;;;;;;;GAYG;AACH,MAAM,IAAI,GAAG,IAAA,kBAAU,EACrB,CAAC,EAA8D,EAAE,GAAG,EAAE,EAAE;QAAvE,EAAE,OAAO,GAAG,MAAM,EAAE,KAAK,GAAG,OAAO,EAAE,IAAI,GAAG,OAAO,OAAW,EAAN,IAAI,cAA5D,4BAA8D,CAAF;IAC3D,OAAO,CACL,8BAAC,wBAAc,oBACT,IAAI,IACR,EAAE,EAAC,GAAG,EACN,OAAO,EAAE,OAAO,EAChB,KAAK,EAAE,KAAK,EACZ,GAAG,EAAE,GAAG,EACR,IAAI,EAAE,IAAI,IACV,CACH,CAAC;AACJ,CAAC,CACF,CAAC;AAEF,IAAI,CAAC,WAAW,GAAG,MAAM,CAAC;AAE1B,kBAAe,IAAI,CAAC"}