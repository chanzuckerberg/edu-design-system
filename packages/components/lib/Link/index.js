"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = void 0;
var Link_1 = require("./Link");
Object.defineProperty(exports, "default", { enumerable: true, get: function () { return __importDefault(Link_1).default; } });
//# sourceMappingURL=index.js.map