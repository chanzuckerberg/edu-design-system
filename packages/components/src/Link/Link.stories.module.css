.color {
  @apply p-2;
}

.variant {
  @apply mb-5;
}

.icon {
  @apply ml-2;
}
