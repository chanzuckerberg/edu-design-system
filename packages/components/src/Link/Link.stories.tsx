import { StoryObj } from "@storybook/react";
import React from "react";
import Text from "../Text";
import {
  colors,
  getStandardSet,
  getPlainRecommendedVariants,
  getLinkRecommendedVariants,
  getDestructiveRecommendedVariants,
  getAllRecommendedVariants,
  getAllVariantsWithStates,
  getLargeVariantsOnDarkBackgroundWithStates,
} from "../storyUtils/clickableStyleUtils";
import Link from "./Link";

export default {
  title: "Link",
  component: Link,
  argTypes: {
    children: {
      control: {
        type: "text",
      },
    },
    color: {
      control: {
        type: "radio",
        options: colors,
      },
    },
  },
  args: {
    children: "Link",
    variant: "link" as const,
    color: "brand" as const,
    href: "",
  },
};

type Args = React.ComponentProps<typeof Link>;

export const Default: StoryObj<Args> = {};

export const LinkInBody: StoryObj<Args> = {
  render: (args) => (
    <Text size="body">
      This text surrounds the <Link {...args} /> and shows that the link should
      adhere to its appearance
    </Text>
  ),
};

export const LinkInHeading: StoryObj<Args> = {
  render: (args) => (
    <Text size="h1">
      This text surrounds the <Link {...args} /> and shows that the link should
      adhere to its appearance
    </Text>
  ),
};

export const PrimaryRecommendedVariants = {
  render: () => getStandardSet(Link, "Link", "flat"),
};

export const SecondaryRecommendedVariants = {
  render: () => getStandardSet(Link, "Link", "outline"),
};

export const TertiaryRecommendedVariants = {
  render: () => getStandardSet(Link, "Link", "outline", "neutral"),
};

export const PlainRecommendedVariants = {
  render: () => getPlainRecommendedVariants(Link, "Link"),
};

export const LinkRecommendedVariants = {
  render: () => getLinkRecommendedVariants(Link, "Link"),
};

export const DestructiveRecommendedVariants = {
  render: () => getDestructiveRecommendedVariants(Link, "Link"),
};

const gridParameters = {
  axe: {
    skip: true,
  },
  // Skip snapshots because this grid has too many components with all the states,
  // and the previous stories generate enough for our needs.
  // These stories are more for visual regression testing.
  snapshot: {
    skip: true,
  },
};

export const AllVariants = {
  render: () => getAllVariantsWithStates("a", "Link"),
  parameters: gridParameters,
};

export const LargeVariantsOnDarkBackground = {
  render: () => getLargeVariantsOnDarkBackgroundWithStates("a", "Link"),
  parameters: {
    ...gridParameters,
    backgrounds: {
      default: "dark",
    },
  },
};

export const AllRecommendedVariants = {
  render: () => getAllRecommendedVariants(Link, "Link"),
};
