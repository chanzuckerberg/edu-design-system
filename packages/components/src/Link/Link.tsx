import React, { ReactNode, forwardRef } from "react";
import Clickable, { ClickableProps } from "../Clickable";

type LinkHTMLElementProps = Omit<
  React.AnchorHTMLAttributes<HTMLAnchorElement>,
  "disabled"
>;

export type LinkProps = LinkHTMLElementProps & {
  as?: "a" | React.ComponentType<any>;
  /**
   * The link contents or label.
   */
  children: ReactNode;
  color?: ClickableProps<"a">["color"];
  "data-testid"?: string;
  size?: ClickableProps<"a">["size"];
  variant?: ClickableProps<"a">["variant"];
};

const Link = forwardRef<HTMLAnchorElement, LinkProps>(
  ({ as = "a", variant = "link", color = "brand", ...rest }, ref) => {
    return (
      <Clickable {...rest} as={as} variant={variant} color={color} ref={ref} />
    );
  },
);

Link.displayName = "Link";

export default Link;
