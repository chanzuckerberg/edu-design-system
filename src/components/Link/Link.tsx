import React, { ReactNode, forwardRef } from "react";
import ClickableStyle, { ClickableStyleProps } from "../ClickableStyle";

type LinkHTMLElementProps = Omit<
  React.AnchorHTMLAttributes<HTMLAnchorElement>,
  "disabled"
>;

export type LinkProps = LinkHTMLElementProps & {
  /**
   * The link contents or label.
   */
  children: ReactNode;
  color?: ClickableStyleProps<"a">["color"];
  "data-testid"?: string;
  size?: ClickableStyleProps<"a">["size"];
  variant?: ClickableStyleProps<"a">["variant"];
};

/**
 * ```ts
 * import {Link} from "@chanzuckerberg/eds";
 * ```
 *
 * Component for making anchor tags.
 *
 * This component is called Link because it should be used to make `<a>` elements;
 * however, it can be styled to look like a button.
 *
 * If you need to style a `<button>` element to look like a link, please use the `Button` component.
 * If you need to style a different element or component (e.g. `Link` from `react-router`) to
 * look like a button or link, you can use the `ClickableStyle` component.
 *
 * In terms of the look and feel of the component in the UI, the `Button`, and `Link`, and `ClickableStyle`
 * components are exactly the same.
 */
const Link = forwardRef<HTMLAnchorElement, LinkProps>(
  ({ variant = "link", color = "brand", size = "large", ...rest }, ref) => {
    return (
      <ClickableStyle
        {...rest}
        as="a"
        variant={variant}
        color={color}
        ref={ref}
        size={size}
      />
    );
  },
);

Link.displayName = "Link";

export default Link;
