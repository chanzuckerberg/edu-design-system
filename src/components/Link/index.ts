export { Link as default } from './Link';
export type { LinkProps } from './Link';
