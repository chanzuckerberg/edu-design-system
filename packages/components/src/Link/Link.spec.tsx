import { generateSnapshots, prepareStory } from "@chanzuckerberg/story-utils";
import { render, screen } from "@testing-library/react";
import React from "react";
import * as LinkStoryFile from "./Link.stories";

describe("<Link />", () => {
  generateSnapshots(LinkStoryFile);

  it("forwards refs", () => {
    const ref = React.createRef<HTMLAnchorElement>();
    render(prepareStory(LinkStoryFile.StandardLink, { ref }));

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    ref.current!.focus();

    const Link = screen.getByRole("link");
    expect(Link).toHaveFocus();
  });
});
