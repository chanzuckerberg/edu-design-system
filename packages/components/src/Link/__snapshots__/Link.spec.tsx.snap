// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> StandardLink story renders snapshot 1`] = `
<a
  class="button variantLink colorBrand"
  href=""
>
  ​
  Link
</a>
`;

exports[`<Link /> buttonVariantWithIcon story renders snapshot 1`] = `
<a
  class="button variantFlat colorBrand"
  href=""
>
  ​
  Link 
  <svg
    aria-hidden="true"
    class="svgIcon"
    fill="currentColor"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zM9.29 16.29L5.7 12.7a.9959.9959 0 010-1.41c.39-.39 1.02-.39 1.41 0L10 14.17l6.88-6.88c.39-.39 1.02-.39 1.41 0 .39.39.39 1.02 0 1.41l-7.59 7.59c-.38.39-1.02.39-1.41 0z"
    />
  </svg>
</a>
`;

exports[`<Link /> linkInBody story renders snapshot 1`] = `
<p
  class="typography sizeBody colorBase"
>
  This text surrounds the 
  <a
    class="button variantLink colorBrand"
    href=""
  >
    ​
    Link
  </a>
   and shows that the link should adhere to its appearance
</p>
`;

exports[`<Link /> linkInHeading story renders snapshot 1`] = `
<p
  class="typography sizeH1 colorBase"
>
  This text surrounds the 
  <a
    class="button variantLink colorBrand"
    href=""
  >
    ​
    Link
  </a>
   and shows that the link should adhere to its appearance
</p>
`;

exports[`<Link /> linkWithIcon story renders snapshot 1`] = `
<a
  class="button variantLink colorBrand"
  href=""
>
  ​
  Link 
  <svg
    aria-hidden="true"
    class="svgIcon"
    fill="currentColor"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zM9.29 16.29L5.7 12.7a.9959.9959 0 010-1.41c.39-.39 1.02-.39 1.41 0L10 14.17l6.88-6.88c.39-.39 1.02-.39 1.41 0 .39.39.39 1.02 0 1.41l-7.59 7.59c-.38.39-1.02.39-1.41 0z"
    />
  </svg>
</a>
`;

exports[`<Link /> withClassName story renders snapshot 1`] = `
<a
  class="example-className button variantLink colorBrand"
  href=""
>
  ​
  Link
</a>
`;

exports[`<Link /> withDataTestId story renders snapshot 1`] = `
<a
  class="button variantLink colorBrand"
  data-testid="fake-test-id"
  href=""
>
  ​
  Link
</a>
`;
